### ISSUE TITLE
Discussion: proper text shaping & layout for consoles or character grids

### ISSUE THREAD

**Issue by be5invis** [OpenType/opentype-layout#11]
This is a discussion thread about how to properly support complex script in a console or a character grid.

Cases may get involved:
![image](https://user-images.githubusercontent.com/240091/55543414-964e3a80-56fb-11e9-8732-d76c15ec1051.png)

**Comment by mhosken** [comment]
I agree that it would be good to have a table of monowidths for characters. I can imagine values from -1 (or even -2) to 2 (or more?) Where negative means the character inserts before other characters and 0 for traditional diacritics. Characters like U+1000 would have width 2 and so on. BTW The example myanmar cluster is not real (thankfully). The width of a cluster should be the same as the sum of the absolute widths of the characters in it.

On further (edited) thought, the negative value is for characters that add width to an existing cluster. So in effect we are also specifying a clustering algorithm at the same time. If the value is > 0 then it is a new cluster. If it is <=0 then it is part of an existing cluster.

Clearly a font would have to be trusted to fit within the space specified.

**Comment by be5invis** [comment]
@mhosken 
From my personal view, text layout on consoles is more like a weird case of justification.
Also, most fonts for east asian / complex script are not monospace, even if they are, they may not share the width of the Latin fonts. For example, CJK fonts use width 1 EM for all EA glyphs, but Latin monospaces fonts are often wider than 1/2 EM, which means that if we fit one EA character into two Latin cells, we need to insert extra space around it.

**Comment by mhosken** [comment]
I think the bigger issue for fixed spacing is application support. If apps
can lookup how far to move a cursor every arrow key, then what the font
does is less important.

**Comment by stone-zeng** [comment]
I think math can also be considered as a kind of complex script and there is actually [such demand](https://github.com/firamath/firamath/issues/20#issuecomment-437653835) to support monospaced math font.

**Comment by davelab6** [comment]
cc @raphlinus
